File: algorithms/greedy/HuffmanCoding.cpp
#include <bits/stdc++.h>
using namespace std;

class Node {
public:
  char ch;
  int freq;
  Node *left;
  Node *right;

  Node(char ch, int freq) {
    this->ch = ch;
    this->freq = freq;
    left = nullptr;
    right = nullptr;
  }

  Node(char ch, int freq, Node *left, Node *right) {
    this->ch = ch;
    this->freq = freq;
    this->left = left;
    this->right = right;
  }
};

struct compare {
  bool operator()(Node *l, Node *r) { return l->freq > r->freq; }
};

void printCodes(Node *root, string str,
                unordered_map<char, string> &huffmanCode) {
  if (root == nullptr) {
    return;
  }

  if (!root->left && !root->right) {
    huffmanCode[root->ch] = str;
  }

  printCodes(root->left, str + "0", huffmanCode);
  printCodes(root->right, str + "1", huffmanCode);
}

void buildHuffmanTree(string text) {
  unordered_map<char, int> freq;

  for (char ch : text) {
    freq[ch]++;
  }

  priority_queue<Node *, vector<Node *>, compare> pq;

  for (auto pair : freq) {
    pq.push(new Node(pair.first, pair.second));
  }

  while (pq.size() != 1) {
    Node *left = pq.top();
    pq.pop();
    Node *right = pq.top();
    pq.pop();

    int sum = left->freq + right->freq;

    pq.push(new Node('\0', sum, left, right));
  }

  Node *root = pq.top();
  unordered_map<char, string> huffmanCode;

  printCodes(root, "", huffmanCode);
  cout << "Huffman Codes:\n";

  for (auto pair : huffmanCode) {
    cout << pair.first << " " << pair.second << "\n";
  }

  cout << "\nOriginal string:\n" << text << "\n";

  string str = "";

  for (char ch : text) {
    str += huffmanCode[ch];
  }

  cout << "\nEncoded string:\n" << str << "\n";

  auto decode = [&](string str) {
    cout << "\nDecoded string:\n";

    Node *curr = root;

    for (char bit : str) {
      if (bit == '0') {
        curr = curr->left;
      } else {
        curr = curr->right;
      }

      if (!curr->left && !curr->right) {
        cout << curr->ch;
        curr = root;
      }
    }

    cout << "\n";
  };

  decode(str);
}

File: algorithms/greedy/HuffmanCoding.java
import java.util.*;

class Node {
  char ch;
  int freq;
  Node left;
  Node right;

  Node(char ch, int freq) {
    this.ch = ch;
    this.freq = freq;
    this.left = null;
    this.right = null;
  }

  Node(char ch, int freq, Node left, Node right) {
    this.ch = ch;
    this.freq = freq;
    this.left = left;
    this.right = right;
  }
}

public class HuffmanCoding {
  public static void printCodes(
      Node root, String str, Map<Character, String> huffmanCode) {
    if (root == null)
      return;

    if (root.left == null && root.right == null) {
      huffmanCode.put(root.ch, str);
    }

    printCodes(root.left, str + "0", huffmanCode);
    printCodes(root.right, str + "1", huffmanCode);
  }

  public static void buildHuffmanTree(String text) {
    Map<Character, Integer> freq = new HashMap<>();
    for (char ch : text.toCharArray()) {
      freq.put(ch, freq.getOrDefault(ch, 0) + 1);
    }

    PriorityQueue<Node> pq = new PriorityQueue<>((l, r) -> l.freq - r.freq);

    for (Map.Entry<Character, Integer> entry : freq.entrySet()) {
      pq.add(new Node(entry.getKey(), entry.getValue()));
    }

    while (pq.size() != 1) {
      Node left = pq.poll();
      Node right = pq.poll();

      int sum = left.freq + right.freq;

      pq.add(new Node('\0', sum, left, right));
    }

    Node root = pq.peek();

    Map<Character, String> huffmanCode = new HashMap<>();
    printCodes(root, "", huffmanCode);

    System.out.println("Huffman Codes:");
    for (Map.Entry<Character, String> entry : huffmanCode.entrySet()) {
      System.out.println(entry.getKey() + " " + entry.getValue());
    }

    System.out.println("\nOriginal string:");
    System.out.println(text);

    StringBuilder encodedStr = new StringBuilder();
    for (char ch : text.toCharArray()) {
      encodedStr.append(huffmanCode.get(ch));
    }

    System.out.println("\nEncoded string:");
    System.out.println(encodedStr);

    System.out.println("\nDecoded string:");
    Node curr = root;

    for (char bit : encodedStr.toString().toCharArray()) {
      if (bit == '0') {
        curr = curr.left;
      } else {
        curr = curr.right;
      }

      if (curr.left == null && curr.right == null) {
        System.out.print(curr.ch);
        curr = root;
      }
    }

    System.out.println();
  }
}
